File: code/backend/python/app/core/ports/auth/token_persistence_port.py
from abc import ABC, abstractmethod
from app.core.models.auth.token_info import TokenInfo
from typing import Optional

class TokenPersistencePort(ABC):
    @abstractmethod
    def save_refresh_token(self, user_uuid: str, refresh_token: str, token_info: TokenInfo) -> None:
        pass

    @abstractmethod
    def revoke_token(self, refresh_token: str) -> None:
        pass

    @abstractmethod
    def revoke_all_by_user_uuid(self, user_uuid: str) -> None:
        pass

    @abstractmethod
    def is_refresh_token_valid(self, refresh_token: str) -> bool:
        pass

    @abstractmethod
    def find_user_id_by_uuid(self, user_uuid: str) -> Optional[int]:
        pass

    @abstractmethod
    def enforce_token_limit(self, user_id: int, max_tokens: int) -> None:
        pass

File: code/backend/php/src/Core/Port/Auth/TokenPersistencePort.php
<?php

namespace Games\Paths\Core\Port\Auth;

use Games\Paths\Core\Domain\Auth\TokenInfo;

/**
 * Outbound port for refresh token persistence in DB.
 */
interface TokenPersistencePort
{
    /**
     * Stores a new refresh token.
     */
    public function saveRefreshToken(string $userUuid, string $refreshToken, TokenInfo $tokenInfo): void;

    /**
     * Revokes a specific refresh token by its JTI or content.
     */
    public function revokeToken(string $refreshToken): void;

    /**
     * Revokes all active refresh tokens for a user.
     */
    public function revokeAllByUserUuid(string $userUuid): void;

    /**
     * Checks if a refresh token exists and is valid (not revoked and not expired).
     */
    public function isRefreshTokenValid(string $refreshToken): bool;

    /**
     * Finds the user unique ID (persistence id) by user UUID.
     */
    public function findUserIdByUuid(string $userUuid): ?int;

    /**
     * Enforces the maximum number of active tokens per user (deletes oldest).
     */
    public function enforceTokenLimit(int $userId, int $maxTokens): void;
}
